src/grid/grid_task_list.c

```diff
@@ -169,27 +251,32 @@ void grid_collocate_task_list(
       grid_ref[level] = malloc(sizeof_grid);
       memset(grid_ref[level], 0, sizeof_grid);
     }
-    // Copy blocks if nessecary (by slightly violating the Law of Demeter).
-    if (task_list->backend == GRID_BACKEND_CPU) {
-      memcpy(task_list->ref->blocks_buffer,
-             task_list->cpu->crutch->blocks_buffer,
-             task_list->ref->buffer_size * sizeof(double));
+
+    switch (task_list->backend) {
+    case GRID_BACKEND_CPU:
+      extract_grid_context_block_buffer(task_list->cpu,
+                                        task_list->ref->blocks_buffer);
+      break;
 #ifdef __GRID_CUDA
-    } else if (task_list->backend == GRID_BACKEND_GPU) {
+    case GRID_BACKEND_GPU:
       memcpy(task_list->ref->blocks_buffer, task_list->gpu->blocks_buffer_host,
              task_list->ref->buffer_size * sizeof(double));
-#endif // __GRID_CUDA
-    } else if (task_list->backend != GRID_BACKEND_REF) {
-      printf("Error: Unknown grid backend: %i.\n", task_list->backend);
-      abort();
+      break;
+    case GRID_BACKEND_HYBRID:
+      extract_grid_context_block_buffer(task_list->hybrid,
+                                        task_list->ref->blocks_buffer);
+      break;
+#endif
+    default:
+      break;
     }
     // Call reference implementation.
     grid_ref_collocate_task_list(task_list->ref, orthorhombic, func, nlevels,
                                  npts_global, npts_local, shift_local,
                                  border_width, dh, dh_inv, grid_ref);
 
     // Compare results.
-    const double tolerance = 1e-12;
+    const double tolerance = 1e-12; // TODO: tune to a reasonable value.
```

Comment:
I actually think 1e-12 is a reasonable value.

Reply:
I have no problem with this.

Reply:
trailing comment. removed.